src/libs/Navigation/AppNavigator/AuthScreens.js

```diff
@@ -352,9 +321,4 @@ AuthScreens.propTypes = propTypes;
 export default compose(
     withWindowDimensions,
     withNetwork(),
-    withOnyx({
-        session: {
-            key: ONYXKEYS.SESSION,
-        },
-    }),
```

Comment:
**explanation:** `withOnyx()` will not render the component until it has at least read the value (could still be null but we know that it has been read into memory by the time the logic in `componentDidMount` runs).

Reply:
I think "weird things" were happening using `Onyx.connect`. I found that `setUpPoliciesAndNavigate` was still running after the `AuthScreens` unmounted and `Policy_Create` was getting called before the `Create_Login` api call finished.

You kind of lost me there @tgolen, so I will pass the session through to the action. I think that's the most straight forward.

Reply:
>But are you saying that it's then OK to remove this subscription, or that it needs to be kept here?

**short answer:** I think if we leave the subscription and pass `session` as a variable that will guarantee it is the value read from storage. But there could be other solutions - leaving the subscription alone and passing as an argument seems simplest for now.

----
**longer answer:**

>so that the Onyx.connect() callback only runs once when the initial value is read (this is done by having an immediate Onyx.disconnect() at the top of the callback).

I think I understand the proposal, but it feels like it's missing something about how to handle the case where a user gets logged out and logged back in (i.e. `setUpPoliciesAndNavigate()` may need to run on every sign in and not just "once"). 

I can see us moving all the logic that needs to run on "sign in" into a big "sign in" event handler and then having an `Onyx.connect()` determine whether we are signed in or not instead of splitting things up in the `AuthScreens` lifecycle methods.

```js
let isLoggedIn = false;
Onyx.connect({
    key: ONYXKEYS.SESSION.
    callback: (val) => {
        // No change
        if ((isLoggedIn && val.authToken) || (!isLoggedIn && !val.authToken)) {
            return;
        }
        if (!isLoggedIn && val.authToken) {
            // run post sign in logic
            isLoggedIn = true;
        } else {
            // run post sign out logic
            isLoggedIn = false;
        }
    },
})
```

The only concern I have with something like that is that it would run before the views of the app, navigation, etc have loaded - but that might be fine.

Reply:
> I think that's the most straight forward

> leaving the subscription alone and passing as an argument seems simplest for now.

I can definitely understand and appreciate both of these arguments, but I still can't shake the feeling that this is an anti-pattern. Let's turn the table around and look at it a different way. If we made a rule out of the logic being applied here, I imagine it looking like this:

- If an action needs access to data in Onyx, it should receive that data from a mounted view

I think we want to avoid that at ALL COST. This is the reason that `Onyx.connect()` exists in the first place. If it's acting "weird" or it's not fitting this flow properly, then I think there are several better approaches:

1. Figure out if there is a bug in `Onyx.connect()` and fix it
2. Add functionality that is missing from `Onyx.connect()`
3. Re-think the entire design and flow of what's happening in this PR

By just not understanding what's happening, and then crafting easy solutions (which are work-arounds and anti-patterns), we introduce patterns into the code which shouldn't be there, and they will get copied and reused by others.

Reply:
> * If an action needs access to data in Onyx, it should receive that data from a mounted view

Yeah I take your point that it's weird to have this "rule". I used `Onyx.connect` initially because my understanding was that non-view code should use it. It did create some funky behavior when I tested it.  This change is what I used to fix the strange behavior but I didn't do a good job of writing down exactly what happened. https://github.com/Expensify/App/blob/35b57776994b1f8efbdc099b55afe85bc6c388ab/src/libs/actions/App.js#L156-L158

There is currently an issue with the Network code. Once that is fixed I will go back to the `Onyx.connect` implementation and I will debug and write down exactly what happens.